kafka.go

```diff
@@ -14,148 +14,254 @@
+	"github.com/segmentio/kafka-go"
+)
+
+const (
+	defaultReaderSize = 10e6 // 10MB
+	maxBackoffCount   = 3
 )
 
-// KafkaClient satisfies the publisher to Kafka
+// KafkaClient wraps client's functionality for Kafka
 type KafkaClient struct {
-	realm         string
-	asyncProducer sarama.AsyncProducer
-	syncProducer  sarama.SyncProducer
+
+	// client realm
+	realm string
+
+	// publish configuration
+	publishConfig kafka.WriterConfig
+
+	// subscribe configuration
+	subscribeConfig kafka.ReaderConfig
+
+	// map to store callback function
+	subscribeMap *sync.Map
 }
 
-// KafkaConfig is Kafka configuration to wait for a successful metadata response
+// KafkaConfig is Kafka configuration to wait dial connection, read and write process
 type KafkaConfig struct {
-	DialTimeout          time.Duration
-	ReadTimeout          time.Duration
-	WriteTimeout         time.Duration
-	MetadataRetryMax     int
-	MetadataRetryBackoff time.Duration
+	DialTimeout  time.Duration
+	ReadTimeout  time.Duration
+	WriteTimeout time.Duration
 }
 
-// NewKafkaClient creates new client to publish event to kafka
-// The first producerConfiguration is the asynchronous producer configuration
-// meanwhile the second is the synchronous one
-func NewKafkaClient(realm string, brokerList []string, producerConfiguration ...*KafkaConfig) (*KafkaClient, error) {
-
-	var configAsync, configSync *sarama.Config
-	switch {
-	case len(producerConfiguration) > 1:
-		configAsync = configureKafka(producerConfiguration[0])
-		configSync = configureKafka(producerConfiguration[1])
-	case len(producerConfiguration) == 1:
-		configAsync = configureKafka(producerConfiguration[0])
-		configSync = sarama.NewConfig()
-	default:
-		configAsync = sarama.NewConfig()
-		configSync = sarama.NewConfig()
-	}
-
-	asyncProducer, err := sarama.NewAsyncProducer(brokerList, configAsync)
-	if err != nil {
-		return nil, err
+func setConfig(writerConfig *kafka.WriterConfig, readerConfig *kafka.ReaderConfig, config *KafkaConfig) {
+
+	if config.ReadTimeout != 0 {
+		writerConfig.ReadTimeout = config.WriteTimeout
 	}
+	if config.WriteTimeout != 0 {
+		writerConfig.WriteTimeout = config.WriteTimeout
+	}
+	if config.DialTimeout != 0 {
+		dialer := &kafka.Dialer{
+			Timeout: config.DialTimeout,
+		}
+		writerConfig.Dialer = dialer
+		readerConfig.Dialer = dialer
+	}
+}
 
-	configSync.Producer.Return.Successes = true
-	syncProducer, err := sarama.NewSyncProducer(brokerList, configSync)
-	if err != nil {
-		return nil, err
+// NewKafkaClient create a new instance of KafkaClient
+func NewKafkaClient(brokers []string, realm string, config ...*KafkaConfig) (*KafkaClient, error) {
+
+	writerConfig := &kafka.WriterConfig{
+		Brokers:  brokers,
+		Balancer: &kafka.LeastBytes{},
 	}
 
-	go listenToError(asyncProducer)
+	readerConfig := &kafka.ReaderConfig{
+		Brokers:  brokers,
+		MaxBytes: defaultReaderSize,
+	}
+
+	// set client configuration
+	if len(config) > 0 {
+		setConfig(writerConfig, readerConfig, config[0])
```

Comment:
either explain that you will only read the first of KafkaConfig array, or handle multiple configs

Reply:
I will add a comment and also update description in the readme
thanks